From a589c0aaf14a34479f417956f6271acbe4f0a3d2 Mon Sep 17 00:00:00 2001
From: Carlos Soriano <csoriano@redhat.com>
Date: Tue, 11 Dec 2018 19:24:36 +0100
Subject: [PATCH 48/55] desktopManager: Remove items from the grid in an idle
 too

So we make sure allocations are properly set and we don't enter an
allocation cycle in case we need to do a new calculation because of an
allocation change in the grid.
---
 desktopManager.js | 24 ++++++++++++++++++------
 1 file changed, 18 insertions(+), 6 deletions(-)

diff --git a/desktopManager.js b/desktopManager.js
index 8f10f19..9175672 100644
--- a/desktopManager.js
+++ b/desktopManager.js
@@ -69,6 +69,7 @@ var DesktopManager = GObject.registerClass({
         super._init(params);
 
         this._layoutChildrenId = 0;
+        this._deleteChildrenId = 0;
         this._scheduleDesktopsRefreshId = 0;
         this._monitorDesktopDir = null;
         this._desktopMonitorCancellable = null;
@@ -553,20 +554,27 @@ var DesktopManager = GObject.registerClass({
         return true;
     }
 
-    _scheduleLayoutChildren() {
-        if (this._layoutChildrenId != 0)
-            return;
+    _resetGridsAndScheduleLayout() {
+        this._deleteChildrenId = 0;
+
+        Object.values(this._desktopGrids).forEach((grid) => grid.reset());
 
         this._layoutChildrenId = GLib.idle_add(GLib.PRIORITY_LOW, () => this._layoutChildren());
+
+        return GLib.SOURCE_REMOVE;
     }
 
     scheduleReLayoutChildren() {
-        if (this._layoutChildrenId != 0)
+        if (this._deleteChildrenId != 0)
             return;
 
-        Object.values(this._desktopGrids).forEach((grid) => grid.reset());
+        if (this._layoutChildrenId != 0) {
+            GLib.source_remove(this._layoutChildrenId);
+            this._layoutChildrenId = 0;
+        }
 
-        this._layoutChildrenId = GLib.idle_add(GLib.PRIORITY_LOW, () => this._layoutChildren());
+
+        this._deleteChildrenId = GLib.idle_add(GLib.PRIORITY_LOW, () => this._resetGridsAndScheduleLayout());
     }
 
     _addFileItemCloseTo(item) {
@@ -685,6 +693,10 @@ var DesktopManager = GObject.registerClass({
             GLib.source_remove(this._layoutChildrenId);
         this._layoutChildrenId = 0;
 
+        if (this._destroyChildrenId)
+            GLib.source_remove(this._destroyChildrenId);
+        this._destroyChildrenId = 0;
+
         if (this._monitorsChangedId)
             Main.layoutManager.disconnect(this._monitorsChangedId);
         this._monitorsChangedId = 0;
-- 
2.20.0

